### ISSUE TITLE
get_base_conf_dir has problems with staging

### ISSUE THREAD

**Issue by arrowd** [Xpra-org/xpra#4056]
When building Xpra package on FreeBSD which uses staging internally I see

```
get_conf_dir(/wrkdirs/usr/ports/x11/xpra/work/stage)=/usr/ports/x11/xpra/work/stage/etc/xpra
```

which is wrong and should be `/usr/ports/x11/xpra/work/stage//usr/local/etc/xpra`. The command I use to install is

```
install --prefix=/usr/local --root=/usr/local/poudriere/ports/default/x11/xpra/work/stage ...
```

**Comment by totaam** [comment]
No idea why this is wrong without more details.

The script needs both the file location in full absolute path form, and also stripped from the buildroot for generating runtime paths.

**Comment by arrowd** [comment]
Could it be because I call `build` with there arguments beforehand?

**Comment by totaam** [comment]
> Could it be because I call `build` with there arguments beforehand?

I don't understand what that means.
You're still not telling what the actual problem is.

My guess is that we need to add a BSDism in here:
https://github.com/Xpra-org/xpra/blob/816b73087e1a7c7c36c1f1cf0574c47409c64618/setup.py#L934-L940

**Comment by arrowd** [comment]
My install dir is `/usr/local`. The problem is that `get_conf_dir` gets `/wrkdirs/usr/ports/x11/xpra/work/stage` as a value for `install_dir`, which is wrong.

I have a feeling that this line gives me wrong value:

https://github.com/Xpra-org/xpra/blob/816b73087e1a7c7c36c1f1cf0574c47409c64618/setup.py#L1760

**Comment by arrowd** [comment]
I tried to hack it around but then it got even more confusing. The output I want is

```
get_conf_dir(/wrkdirs/usr/ports/x11/xpra/work/stage)=/wrkdirs/usr/ports/x11/xpra/work/stage/usr/local/etc/xpra
```

**Comment by totaam** [comment]
Sorry, but it's still not clear to me what environment variables / settings (`sys.prefix`?) should be used and where.
You're quoting a single line without any context, without explaining what these directories are.

**Comment by arrowd** [comment]
Staging is a technique used by downstream packaging systems when the software gets installed into an arbitrary empty directory (not the prefix) so that packaging system can check what files are actually being installed. In case of plain Makefiles it is often achieved via a variable (usually called `DESTDIR`). When the rule runs as `make install DESTDIR=/stage` the software gets installed inside `/stage` dir but with prefix configured into usual `/usr` or `/usr/local`.

In my case `/wrkdirs/usr/ports/x11/xpra/work/stage` is a staging dir and the prefix is `/usr/local`. I expect the `setup.py` script to use `/wrkdirs/usr/ports/x11/xpra/work/stage/usr/local/etc` as a configuration dir, `/wrkdirs/usr/ports/x11/xpra/work/stage/usr/local/lib/pythonX.Y/site-packages` for Python stuff, etc.

**Comment by totaam** [comment]
> Staging is a technique used by downstream packaging systems...

Yes, the whole point of this function is to detect such build roots / staging directories.
Sounds trivial to fix.

**Comment by arrowd** [comment]
Building with `pkgdir=/wrkdirs/usr/ports/x11/xpra/work/stage/` results in

```
get_conf_dir(/wrkdirs/usr/ports/x11/xpra/work/stage)=/usr/ports/x11/xpra/work/stage/etc/xpra
```

which is not what I want. For reference,

```
install_data_override.run() install_dir=/wrkdirs/usr/ports/x11/xpra/work/stage/usr/local
install_data_override.run() root_prefix=/wrkdirs/usr/ports/x11/xpra/work/stage
```

**Comment by totaam** [comment]
Are you sure you're doing it correctly on a clean copy of git master? Or maybe remove the trailing slash from `pkgdir`.
The resulting conf dir used shows that it probably matched the next rule (strip at `/usr/`):
https://github.com/Xpra-org/xpra/blob/0a21c1ae69af9fa1eb1be964e53c17c2189d1f16/setup.py#L995-L999

**Comment by arrowd** [comment]
I'm building 5.0.3 while applying patches on top of it. With the `lstrip` change and without trailing `/` in the `pkgdir` value I now get:

```
get_conf_dir(/wrkdirs/usr/ports/x11/xpra/work/stage)=/etc/xpra
```

**Comment by totaam** [comment]
> I now get:

Without context, I have no idea if that's correct or not.

**Comment by totaam** [comment]
> What more context do you need?

2 weeks of back and forth and I only just get the command lines now!?

> but now with Xpra 5 it results in wrong file layout

So this is a regression?

You still don't seem to understand what this method does or why it exists, there are docstrings that explain it all.
Without knowing the value of `stripbuildroot`, it is impossible to say if a path starting with `/usr` is correct or not.

I'm done, please file a PR if you work it out.

**Comment by arrowd** [comment]
> 2 weeks of back and forth and I only just get the command lines now!?

Did you explicitly ask for them? I'm not a native English speaker so the word "context" probably confused me.

> So this is a regression?

Yes, it is, the current port for Xpra 4.4.5 arranges the files correctly.

> I'm done, please file a PR if you work it out.
> closed this as completed

I don't quite understand this reaction. If you believe this is not a bug on Xpra side then what kind of PR do you expect? If it is me who build Xpra the wrong way, what should I do? I tried passing various values for `pkgdir` without luck.

Your response feel a bit offensive to me, to be honest. Where was I wrong to deserve this pushback?

**Comment by totaam** [comment]
I have already spent way too much time on this, hence why I will no longer spend time on it.
